internal/agent/device/dependency/dependency.go

```diff
@@ -1,20 +1,406 @@
+	pullTimeout util.Duration,
+) *prefetchManager {
+	return &prefetchManager{
+		log:          log,
+		podmanClient: podmanClient,
+		pullTimeout:  time.Duration(pullTimeout),
+		tasks:        make(map[string]*prefetchTask),
+		queue:        make(chan string, maxQueueSize),
+	}
+}
+
+func (m *prefetchManager) Run(ctx context.Context) {
+	m.log.Debug("Prefetch manager started")
+	defer m.log.Debug("Prefetch manager stopped")
+
+	go m.worker(ctx)
+
+	<-ctx.Done()
+}
+
+func (m *prefetchManager) worker(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Debugf("Prefetch worker exiting: %v", ctx.Err())
+			return
+
+		case image := <-m.queue:
+			m.processTarget(ctx, image)
+		}
+	}
+}
+
+func (m *prefetchManager) RegisterOCICollector(collector OCICollectorFn) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.collectors = append(m.collectors, collector)
+}
+
+func (m *prefetchManager) BeforeUpdate(ctx context.Context, current, desired *v1alpha1.DeviceSpec) error {
+	m.log.Debug("Collecting OCI targets from all registered prefetch functions")
+
+	// collect all OCI targets from registered functions
+	var allTargets []OCIPullTarget
+	m.mu.Lock()
+	collectors := slices.Clone(m.collectors)
+	m.mu.Unlock()
+
+	for i, collector := range collectors {
+		targets, err := collector(ctx, current, desired)
+		if err != nil {
+			return fmt.Errorf("prefetch function %d failed: %w", i, err)
+		}
+		allTargets = append(allTargets, targets...)
+	}
+
+	if len(allTargets) > 0 {
+		m.log.Debugf("Scheduling %d total OCI targets for prefetching", len(allTargets))
+		if err := m.Schedule(ctx, allTargets); err != nil {
+			return fmt.Errorf("scheduling prefetch targets: %w", err)
+		}
+	}
+
+	return m.checkReady(ctx)
+}
+
+func (m *prefetchManager) checkReady(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var pending []string
+	for image, task := range m.tasks {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if !task.done {
+			pending = append(pending, fmt.Sprintf("%s in progress", image))
+			continue
+		}
+		if task.err != nil {
+			if errors.IsRetryable(task.err) {
+				pending = append(pending, fmt.Sprintf("%s retrying: %v", image, task.err))
+			} else {
+				return task.err
+			}
+		}
+	}
+
+	if len(pending) > 0 {
+		return fmt.Errorf("%w: %v", errors.ErrPrefetchNotReady, pending)
+	}
+	return nil
+}
+
+func (m *prefetchManager) processTarget(ctx context.Context, target string) {
+	var task *prefetchTask
+	m.log.Infof("Prefetching OCI target: %s", target)
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	m.mu.Lock()
+	task = m.tasks[target]
+	task.cancelFn = cancel
+	m.mu.Unlock()
+
+	defer cancel()
+
+	for {
+		if ctx.Err() != nil {
+			m.setResult(target, fmt.Errorf("pulling oci target %s: %w", target, ctx.Err()))
+			m.log.Warnf("Context error pulling oci target: %s", target)
+			return
+		}
+		if err := m.pull(ctx, target, task); err != nil {
+			if errors.IsRetryable(err) {
+				time.Sleep(queueNextItemAfter)
```

Comment:
minor: is a:
```go
select {
<-time.After(queueNextItemAfter)
<-ctx.Done()
}
```
preferable here to save those potential sweet sweet seconds of sleeping